### PR TITLE
Typescript (Fix issues and add support for BC)

### PR DESCRIPTION
**Pull request by KiranNiranjan**
## Description
Fix issues and add support for BC

- Auto Launch
- First-time launch
- Basic Auth
- Desktop capturer
- Code documentation
- Network error handler

### Screenshot
Screenshot 2019-04-01 at 1 40 39 PM

## QA Checklist
- [X] Unit-Tests
- [ ] Automation-Tests

### CODE REVIEW

**Comment by KiranNiranjan** [comment]
#### new commit: Added support for certificate verification proxy and whitelisting

**Comment by vrbsm** [comment]
@KiranNiranjan please rebase to fix this issue: `export const BrowserWindow = {`

**Comment by VishwasShashidhar** [comment]
@KiranNiranjan Please rebase against the `typescript-2` branch.

File: demo/index.html [hunk 1]
```diff
@@ -316,7 +302,135 @@ <h1>Symphony Electron API Demo</h1>
...
+         * Activity Detection
+         */
+
+        const activityDetection = document.getElementById('activity-detection');
+        const activityStatus = document.getElementById('activity-status');
+        activityDetection.addEventListener('click', () => {
+            if (window.ssf) {
+                window.ssf.registerActivityDetection(10000, activityDetection);
+            } else {
+                postMessage(apiCmds.registerActivityDetection, 5000);
```
Comment (VishwasShashidhar) [inline demo/index.html]: why is the timer different here from ssf?

File: demo/index.html [hunk 2]
```diff
@@ -316,7 +302,135 @@ <h1>Symphony Electron API Demo</h1>
...
+                        document.getElementById('video').srcObject = null;
+                    }
+                });
+            } else {
+                postMessage(apiCmds.openScreenSharingIndicator, { streamId: stream.id, displayId, requestId: ++requestId });
+            }
+        };
+
+        const handleError = (e) => {
+            console.log(e)
```
Comment (VishwasShashidhar) [inline demo/index.html]: change this to `console.error`

File: demo/index.html [hunk 3]
```diff
@@ -316,7 +302,135 @@ <h1>Symphony Electron API Demo</h1>
...
+            const { data, method } = event.data;
+            switch (method) {
+                case 'activity-callback':
+                    activityCallback(data);
+                    break;
+                case 'media-source-callback':
+
+                    if (data.error) {
+                        if (data.error.message === 'User Cancelled') {
+                            console.log('user canceled');
```
Comment (VishwasShashidhar) [inline demo/index.html]: change this to `console.error`

File: demo/index.html [hunk 4]
```diff
@@ -316,7 +302,135 @@ <h1>Symphony Electron API Demo</h1>
...
+                            chromeMediaSourceId: data.source.id
+                        },
+                        optional: []
+                    };
+                    navigator.mediaDevices.getUserMedia({ video: constraints }).then((stream) => {
+                        handleStream(stream, data.source.display_id);
+                    });
+                    break;
+                case 'screen-sharing-indicator-callback':
+                    if (typeof data.type !== 'string' || !requestId) {
```
Comment (VishwasShashidhar) [inline demo/index.html]: would it be better to add a check to ensure data & type are not undefined?

File: demo/index.html (hunk 1 above)
Comment (KiranNiranjan) [reply demo/index.html]: Both of them works. In `ssf` it is the same as the logic that changes the colour.

**Comment by KiranNiranjan** [comment]
**New commit:** Added network failure error handler screens

**Comment by KiranNiranjan** [comment]
**New commit:** Updated unit test case in `loading screen` & `app menu`

Screenshot 2019-04-01 at 12 46 52 PM

File: src/app/window-handler.ts [hunk 5]
```diff
@@ -259,27 +247,38 @@ export class WindowHandler {
...
-            if (this.isCustomTitleBarAndWindowOS) {
-                this.mainWindow.webContents.send('initiate-custom-title-bar');
-            }
 
             this.mainWindow.webContents.send('page-load', {
                 isWindowsOS,
                 locale: i18n.getLocale(),
                 resources: i18n.loadedResources,
                 origin: this.globalConfig.url,
+                enableCustomTitleBar: this.isCustomTitleBarAndWindowOS,
```
Comment (VishwasShashidhar) [inline src/app/window-handler.ts]: do we need such a big name `isCustomTitleBarAndWindowOS` given that custom title bar is only applicable for windows. Can we change it to just `this.isCustomTitleBar`?

File: src/app/window-handler.ts [hunk 6]
```diff
@@ -259,27 +247,38 @@ export class WindowHandler {
...
+        this.mainWindow.webContents.on('did-fail-load', (_event, errorCode, errorDesc, validatedURL) => {
+            logger.error(`Failed to load ${validatedURL}, with an error: ${errorCode}::${errorDesc}`);
+            this.loadFailError = errorDesc;
+        });
 
-            // Ready to show the window
-            this.mainWindow.show();
+        this.mainWindow.webContents.on('did-stop-loading', () => {
+            if (this.mainWindow && windowExists(this.mainWindow)) {
+            this.mainWindow.webContents.executeJavaScript('document.location.href').then((href) => {
```
Comment (VishwasShashidhar) [inline src/app/window-handler.ts]: will this have any security issues?

Comment (KiranNiranjan) [reply src/app/window-handler.ts]: I don't think so. This block of code is executed by `electron` and the parameter is hardcoded

**Comment by KiranNiranjan** [comment]
**New commit:** Download manager - fixed icon flashing issue and added demo content

**Pull request merged**